ddtrace/internal/processor/trace.py

```diff
@@ -68,6 +68,33 @@ def process_trace(self, trace):
         return None
 
 
+@attr.s
+class TraceTopLevelSpanProcessor(TraceProcessor):
+    """Processor marks spans as top level"""
+
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        """Mark a span in a trace as top level if:
+         1. Span is a local root
+         2. Span has a different service name than it's parent
+
+        Explicitly set top level to zero/false if:
+         1. Span has a truthy parent_id (not zero or None)
+            AND parent span in not in the trace chunk
+        """
+
+        span_ids = {span.span_id for span in trace}
+        for span in trace:
+            if span is span._local_root:
+                span.set_metric("_dd.top_level", 1)
+            elif span._parent and span.service != span._parent.service:
+                span.set_metric("_dd.top_level", 1)
+            elif span.parent_id and span.parent_id not in span_ids:
```

Comment:
Do we have an idea of the performance impact of this trace processor? Would it save us much if we memoised the result of `span.parent_id not in span_ids`?

Reply:
This is an interesting question, I suppose it depends on whether it is a wide or deep trace.

I'd say wide traces (a lot of spans with the same parent id) are fairly rare. Usually it only happens when there is a function that makes a lot of database queries. But for that we are looking at an upper bound of about 20k child spans for the parent. Otherwise the average case is probably much closer to 1, I'd say definitely <10 on average.

Reply:
I see. I think it's still worth being aware of what the cost is, perhaps compared to the other processors, just to have a rough idea. Also, forget about the memoisation, because I just happened to realise that `span_ids` is a `set`, so it's already memoised 😞 .